### PR TITLE
Sysctl reconciler

### PR DESCRIPTION
**Pull request by dylandreimerink**
This PR both modularizes the sysctl package and makes it reconcile. Sysctl logic is now provided via Hive. `sysctl.Sysctl` provides a synchronous API which is a layer atop the sysctl table and reconciler.  Whenever a sysctl property is set the desired state is stored in the table. The reconciler will apply the sysctl and make sure that it stays applied even if external factors change it in the future. The synchronous layer waits for for the reconciler to be done, in case the sysctl is critical for further operations.

Part of #30024

```release-note
Added sysctl setting reconciliation
```

### CODE REVIEW

**Review by pippolo84 — APPROVED** [review]

File: cilium-health/launch/endpoint.go
Comment (gandro) [inline cilium-health/launch/endpoint.go]: The code structure makes it far from obvious (something we ought to fix), but this actually has been running inside cilium-agent for years now. So you can use the helper which uses the reconciler in the background.

File: pkg/datapath/connector/add.go
Comment (gandro) [inline pkg/datapath/connector/add.go:48]: I'm not sure I have actionable feedback here, but there is out of tree code multi-network code which calls this function from the CNI binary. In the end, I guess it is the responsibility of downstream to catch up which those kind of changes.

**Review by gandro — CHANGES_REQUESTED** [review]
Only reviewed my code owners, one change requested.

The "runs in hive" vs. "does not run in Hive" API difference is a bit unfortunate, in particular since the code needs to be made "generic" by passing in the `sysctlDisable` callback. 

I don't think we will ever stop doing sysctls from the CNI binary due to permission issues (i.e. having those calls be done from cilium-agent requires additional Linux capabilities). I think we should make sure that using sysctl commands from there remains easy.

Would it be possible to provide two implementations of the `sysctl.Sysctl` interface, one based on the reconiler, and one based on one-of writes for the CNI plugin? That way, at least we do not have to duplicate the API.

File: pkg/datapath/connector/add.go
Comment (dylandreimerink) [reply pkg/datapath/connector/add.go:48]: If this function is indeed used externally then they will need to change it, but that shouldn't be hard with the `NewDirectSysctl` constructor I added.

File: daemon/cmd/daemon_main.go
Comment (gandro) [inline daemon/cmd/daemon_main.go:1827]: nit: The commit message still mentioned cilium-health using the direct interface, and still mentions `{Read,Write}Sysctl` as methods.

File: pkg/datapath/connector/veth.go
Comment (gandro) [inline pkg/datapath/connector/veth.go]: Why is the callback still necessary? Cant we pass a `sysctl.Sysctl` to `connector.SetupVeth`?

**Review by gandro — COMMENTED** [review]
Thanks for addressing the feedback, I think this looks much better now. A few minor things that stood out to me

File: daemon/cmd/daemon_main.go
Comment (gandro) [inline daemon/cmd/daemon_main.go:1827]: non-blocking nit: Commit message still mentions "IOW where a statedb and the
reconcilers are not available (e.g: cilium-health and cilium-cni)", even though cilium-health does use the reconciler

File: plugins/cilium-docker/driver/driver.go
Comment (gandro) [inline plugins/cilium-docker/driver/driver.go]: This doesn't compile anymore

**Review by gandro — APPROVED** [review]
Looks good to me now (modulo the compile error, which I assume is going to be fixed before merge), thanks a lot.

File: plugins/cilium-docker/driver/driver.go
Comment (dylandreimerink) [reply plugins/cilium-docker/driver/driver.go]: Sorry about that one. Refactoring can be harder than it looks sometimes :smiling_face_with_tear:

File: daemon/cmd/daemon_main.go
Comment (dylandreimerink) [reply daemon/cmd/daemon_main.go:1827]: Right :face_exhaling:

**Review by rgo3 — APPROVED** [review]

**Review by rolinh — APPROVED** [review]
Vendor changes lgtm (afero is apache 2.0 licensed)

**Review by aanm — APPROVED** [review]

**Review by asauber — APPROVED** [review]
cli-related changes look good. the new constants for the table names are appreciated!

File: pkg/datapath/connector/veth.go
Comment (aspsk) [inline pkg/datapath/connector/veth.go]: Why was this change required? Ok to keep it if you think this looks more straightforward, just wondering.

File: pkg/datapath/linux/sysctl/ops.go
Comment (aspsk) [inline pkg/datapath/linux/sysctl/ops.go:34]: Shouldn't we set `changed` to false here? Otherwise this is incorrect:
```
.../sysctl-path = X
Update(.../sysctl-path, X, &changed) // say, changed was true before call
sysctl is not changed, but changed=true for the caller
```

**Review by aspsk — COMMENTED** [review]
Looks great! I've left two minor questions

File: pkg/datapath/connector/veth.go
Comment (dylandreimerink) [reply pkg/datapath/connector/veth.go]: I don't think it is. Seems like a half attempt at removing the exported function, but its still in use :smile:. I can clean that up.

**Review by aspsk — APPROVED** [review]

File: pkg/datapath/linux/sysctl/ops.go
Comment (dylandreimerink) [reply pkg/datapath/linux/sysctl/ops.go:34]: I believe the current code does exactly what the interface describes. If `changed` is non-nil, its false by default and we only need to change it to true if we did make a change. And the function first does a read before we write and quiets before changed=true if there is nothing to do.

**Pull request merged**
